File: frontend/src/api/contract.ts
import request from './request'

export interface ContractItem {
  id: string
  contract_id: string
  item_name: string
  standard: string | null
  audit_days: number | null
  unit_price: number
  quantity: number
  discount: number
  amount: number
  item_type: string | null
  sort_order: number
}

export interface PaymentPlan {
  id: string
  contract_id: string
  installment_no: number
  description: string | null
  plan_amount: number
  due_date: string
  status: string
  created_at: string
}

export interface PaymentPlanListItem {
  id: string
  contract_id: string
  contract_no: string
  contract_name: string
  installment_no: number
  description: string | null
  plan_amount: number
  due_date: string
  status: string
  received_amount: number
  created_at: string
}

export interface PaymentRecord {
  id: string
  plan_id: string
  contract_id: string
  received_amount: number
  received_date: string
  payment_method: string
  bank_reference: string | null
  received_by: string
  remark: string | null
  created_at: string
}

export interface ContractListItem {
  id: string
  contract_no: string
  customer_id: string
  contract_name: string
  contract_type: string
  certification_standard: string
  total_amount: number
  sign_date: string | null
  start_date: string | null
  end_date: string | null
  status: string
  sales_person: string
  created_at: string
}

export interface ContractDetail extends ContractListItem {
  opp_id: string | null
  service_scope: string
  tax_rate: number | null
  remark: string | null
  approved_by: string | null
  created_by: string | null
  updated_at: string
  items: ContractItem[]
  payment_plans: PaymentPlan[]
}

export const contractApi = {
  list(params: { page?: number; page_size?: number; status?: string; keyword?: string; customer_id?: string }) {
    return request.get<any, { code: number; data: ContractListItem[]; total: number }>('/contracts', { params })
  },
  get(id: string) {
    return request.get<any, { code: number; data: ContractDetail }>(`/contracts/${id}`)
  },
  create(data: Record<string, unknown>) {
    return request.post<any, { code: number; data: ContractDetail }>('/contracts', data)
  },
  update(id: string, data: Record<string, unknown>) {
    return request.put<any, { code: number; data: ContractDetail }>(`/contracts/${id}`, data)
  },
  delete(id: string) {
    return request.delete(`/contracts/${id}`)
  },
  listPaymentPlans(contractId: string) {
    return request.get<any, { code: number; data: PaymentPlan[] }>(`/contracts/${contractId}/payment-plans`)
  },
  listPaymentRecords(contractId: string) {
    return request.get<any, { code: number; data: PaymentRecord[] }>(`/contracts/${contractId}/payment-records`)
  },
  addPaymentRecord(contractId: string, data: Record<string, unknown>) {
    return request.post<any, { code: number; data: PaymentRecord }>(`/contracts/${contractId}/payment-records`, data)
  },
  listAllPaymentPlans(params: { page?: number; page_size?: number; status?: string; overdue?: boolean; keyword?: string }) {
    return request.get<any, { code: number; data: PaymentPlanListItem[]; total: number }>('/contracts/payment-plans/all', { params })
  },
}

File: backend/app/schemas/contract.py
import uuid
from datetime import datetime, date
from decimal import Decimal
from pydantic import BaseModel, ConfigDict, field_validator


# ── 合同服务明细 ─────────────────────────────────────────────
class ContractItemBase(BaseModel):
    item_name: str
    standard: str | None = None
    audit_days: Decimal | None = None
    unit_price: Decimal
    quantity: Decimal = Decimal("1")
    discount: Decimal = Decimal("1.0000")
    amount: Decimal
    item_type: str | None = None
    sort_order: int = 0


class ContractItemOut(ContractItemBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    contract_id: uuid.UUID


# ── 付款计划 ─────────────────────────────────────────────────
class PaymentPlanBase(BaseModel):
    installment_no: int
    description: str | None = None
    plan_amount: Decimal
    due_date: date | None = None
    status: str = "待支付"


class PaymentPlanOut(PaymentPlanBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    contract_id: uuid.UUID
    created_at: datetime


# ── 收款记录 ─────────────────────────────────────────────────
class PaymentRecordCreate(BaseModel):
    plan_id: uuid.UUID
    received_amount: Decimal
    received_date: date
    payment_method: str
    bank_reference: str | None = None
    remark: str | None = None


class PaymentRecordOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    plan_id: uuid.UUID
    contract_id: uuid.UUID
    received_amount: Decimal
    received_date: date
    payment_method: str
    bank_reference: str | None
    received_by: uuid.UUID
    remark: str | None
    created_at: datetime


# ── 合同 ─────────────────────────────────────────────────────
class ContractBase(BaseModel):
    customer_id: uuid.UUID
    opp_id: uuid.UUID | None = None
    contract_name: str
    contract_type: str
    certification_standard: str
    service_scope: str
    total_amount: Decimal
    tax_rate: Decimal | None = Decimal("6.00")
    sign_date: date | None = None
    start_date: date | None = None
    end_date: date | None = None
    status: str = "草稿"
    remark: str | None = None
    sales_person: uuid.UUID

    @field_validator("opp_id", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: object) -> object:
        if v == "" or v is None:
            return None
        return v


class ContractCreate(ContractBase):
    items: list[ContractItemBase] = []
    payment_plans: list[PaymentPlanBase] = []


class ContractUpdate(BaseModel):
    contract_name: str | None = None
    contract_type: str | None = None
    certification_standard: str | None = None
    service_scope: str | None = None
    total_amount: Decimal | None = None
    tax_rate: Decimal | None = None
    sign_date: date | None = None
    start_date: date | None = None
    end_date: date | None = None
    status: str | None = None
    remark: str | None = None
    sales_person: uuid.UUID | None = None


class ContractOut(ContractBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    contract_no: str
    approved_by: uuid.UUID | None
    created_by: uuid.UUID | None
    created_at: datetime
    updated_at: datetime
    items: list[ContractItemOut] = []
    payment_plans: list[PaymentPlanOut] = []


class ContractListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    contract_no: str
    customer_id: uuid.UUID
    contract_name: str
    contract_type: str
    certification_standard: str
    total_amount: Decimal
    sign_date: date | None
    start_date: date | None
    end_date: date | None
    status: str
    sales_person: uuid.UUID
    created_at: datetime


# ── 付款计划汇总（全局列表用）────────────────────────────────
class PaymentPlanListItem(BaseModel):
    id: uuid.UUID
    contract_id: uuid.UUID
    contract_no: str
    contract_name: str
    installment_no: int
    description: str | None
    plan_amount: Decimal
    due_date: date | None
    status: str
    received_amount: Decimal  # sum of payment records for this plan
    created_at: datetime
